In [139]:
import pandas as pd
import numpy as np

In [140]:
from sklearn.metrics import mean_squared_error

In [141]:
import os

In [142]:
y_true = pd.read_csv('datasets/test_set.csv')['cpi_pct'].to_numpy()

In [143]:
prediction = []
for file_name in os.listdir('predictions/'):
    
    if file_name[:4] != 'pca5':
        continue
    
    y_pred = pd.read_csv(f'predictions/{file_name}', index_col=0).to_numpy()
    
    name = file_name.rsplit('.', 1)[0]
    rmse =  mean_squared_error(y_true, y_pred, squared=False)
    print(name, "rmse:", round(rmse, 7))
    prediction.append({'model': name, 'rmse': rmse})

pca50_lasso_cv rmse: 0.0022366
pca50_lasso_cvshuffled rmse: 0.0032184
pca50_ols rmse: 0.002213
pca50_ridge_cv rmse: 0.0022573


In [124]:
sorting = ['benchmark1_mean', 'benchmark2_prev', 'benchmark3_rf', 
           'ols','ridge_cv', 'lasso_cv', 'lasso_cvshuffled',
           'lasso_bic', 'lasso_ebic', 'bma']

In [125]:
score_df = pd.DataFrame(prediction).set_index('model')

In [126]:
score_df['diff to best'] =  score_df.rmse - score_df.rmse.min()

In [127]:
#score_df.loc[sorting, :].to_csv('results/1970_2023_model_scores.csv')
score_df.sort_values('rmse').to_csv('results/1970_2023_pca_model_scores_sorted.csv')

In [137]:
sorted_cols = score_df.sort_values('rmse').index

2000-2023 model:

- benchmark1_mean rmse: 0.0026948
- benchmark2_prev rmse: 0.0034082
- bma rmse: 0.002297
- lasso_bic rmse: 0.0023243
- lasso_ebic rmse: 0.0023711
- lasso_cv rmse: 0.0027338
- lasso_cvshuffled rmse: 0.0027338
- ols rmse: 0.0279417
- ridge_cv rmse: 0.0024787

1970-2023 model:

- benchmark1_mean rmse: 0.0032184
- benchmark2_prev rmse: 0.0026672
- benchmark3_rf rmse: 0.0020702
- bma rmse: 0.0020429
- lasso_bic rmse: 0.0021015
- lasso_cv rmse: 0.0021016
- lasso_cvshuffled rmse: 0.0019987
- lasso_ebic rmse: 0.0021058
- ols rmse: 0.0030478
- ridge_cv rmse: 0.0021375

1970-2023 model (PCA):

- benchmark1_mean rmse: 0.0026948
- benchmark2_prev rmse: 0.0034082
- benchmark3_rf rmse: 0.0020702
- bma rmse: 0.002297
- lasso_bic rmse: 0.0023243
- lasso_ebic rmse: 0.0023711
- lasso_cv rmse: 0.0027338
- lasso_cvshuffled rmse: 0.0027338
- ols rmse: 0.0279417
- ridge_cv rmse: 0.0024787

In [129]:
df_list = []


for file_name in os.listdir('models/coefs/'):
    
    coefs = pd.read_csv(f'models/coefs/{file_name}', index_col=0)
    
    if file_name[:4] != 'pca_':
        continue
    
    if file_name in ["bma_coefs.csv", "pca_bma_coefs.csv"]:  
        coefs = coefs.rename({'intercept': 'Intercept'})
     
    name = file_name.rsplit('.', 1)[0]
    
    if name[-5:] != 'coefs':
        continue

    coefs.columns = [name[:-6]]
    

    print(name)
    df_list.append(coefs)


pca_bma_coefs
pca_lasso_bic_coefs
pca_lasso_cvshuffled_coefs
pca_lasso_cv_coefs
pca_lasso_ebic_coefs
pca_ols_coefs
pca_ridge_cv_coefs


In [130]:
df = pd.concat(df_list, axis=1).replace(-0, 0)

In [131]:
df.to_csv('results/all_pca_model_coefs.csv')

In [133]:
sorted_idx = df.abs().sort_values('pca_lasso_cv', ascending=False).index

In [138]:
df.loc[sorted_idx, sorted_cols].to_csv('results/all_pca_model_coefs_sorted.csv')

In [135]:
df.columns

Index(['pca_bma', 'pca_lasso_bic', 'pca_lasso_cvshuffled', 'pca_lasso_cv',
       'pca_lasso_ebic', 'pca_ols', 'pca_ridge_cv'],
      dtype='object')

In [ ]:
['pca_bma', 'pca_lasso_bic', 'pca_lasso_cvshuffled', 'pca_lasso_cv',
       'pca_lasso_ebic', 'pca_ols', 'pca_ridge_cv']